File: my-folder/0152-maximum-product-subarray/solution.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        currMax = 1
        currMin = 1
        max_prod = float("-inf")
        for i in range(len(nums)):
            tmp = nums[i] * currMax
            currMax = max(tmp, currMin * nums[i], nums[i])
            currMin = min(tmp, currMin * nums[i], nums[i])
            
            max_prod = max(currMax, max_prod)
        
        return max_prod

        

File: my-folder/0152-maximum-product-subarray/solution.java
class Solution {
    public int maxProduct(int[] nums) {
        int currMin = 1;
        int currMax = 1;
        int maxProduct = Integer.MIN_VALUE;
        for(int num : nums){
            int temp = num * currMax;
            currMax = Math.max(num, Math.max(temp, currMin * num));
            currMin = Math.min(num, Math.min(temp, currMin * num));

            maxProduct = Math.max(maxProduct, currMax);
        }

        return maxProduct;   
    }
}

